tensorflow-framework/src/main/java/org/tensorflow/framework/activations/ReLU.java

```diff
@@ -0,0 +1,143 @@
+ *     relu = new ReLU&lt;&gt;(tf, ReLU.ALPHA_DEFAULT, ReLU.MAX_VALUE_DEFAULT, 5f);
+ *     result = relu.call(input);
+ *     // result is [-0.f, -0.f,  0.f,  0.f, 10.f]
+ * </pre>
+ *
+ * @param <T> the data type of the result
+ */
+public class ReLU<T extends TNumber> extends Activation<T> {
+
+  public static final float ALPHA_DEFAULT = 0.0f;
+  public static final float MAX_VALUE_DEFAULT = Float.NaN;
+  public static final float THRESHOLD_DEFAULT = 0.0f;
+
+  private final float alpha;
+  private final float maxValue;
+  private final float threshold;
+
+  /**
+   * Creates a new ReLU with alpha={@link #ALPHA_DEFAULT}, maxValue={@link #MAX_VALUE_DEFAULT},
+   * threshold={@link #THRESHOLD_DEFAULT},
+   *
+   * @param tf the TensorFlow Ops
+   */
+  public ReLU(Ops tf) {
+    this(tf, ALPHA_DEFAULT, MAX_VALUE_DEFAULT, THRESHOLD_DEFAULT);
+  }
+
+  /**
+   * Creates a new ReLU
+   *
+   * @param tf the TensorFlow Ops
+   * @param alpha governs the slope for values lower than the threshold.
+   * @param maxValue sets the saturation threshold (the largest value the function will return).
+   * @param threshold the threshold value of the activation function below which values will be
+   *     damped or set to zero.
+   */
+  public ReLU(Ops tf, float alpha, float maxValue, float threshold) {
+    super(tf);
+    this.alpha = alpha;
+    this.maxValue = maxValue;
+    this.threshold = threshold;
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public Operand<T> call(Operand<T> input) {
+
+    DataType<T> dataType = input.asOutput().dataType();
+    Operand<T> negativePart = null;
+    boolean clipMax = !Float.isNaN(maxValue);
+    if (alpha != 0) {
+      if (Float.isNaN(maxValue) && threshold == 0) {
+        // TODO LeakyRelu is not in tf.nn ????
+        return LeakyRelu.create(tf.scope(), input, LeakyRelu.alpha(alpha));
+      }
+      if (threshold != 0) {
+        negativePart =
+            tf.nn.relu(
+                tf.math.add(tf.math.neg(input), tf.dtypes.cast(tf.constant(threshold), dataType)));
+      } else {
+        negativePart = tf.nn.relu(tf.math.neg(input));
+      }
+    }
+
+    if (threshold != 0) {
+      // computes input for input > threshold else 0
+      Greater greater = tf.math.greater(input, tf.dtypes.cast(tf.constant(threshold), dataType));
+      input = tf.math.mul(input, tf.dtypes.cast(greater, dataType));
+    } else if (maxValue == 6) {
+      // if no threshold, then can use nn.relu6 native TF op for performance
+      input = tf.nn.relu6(input);
+      clipMax = false;
+    } else {
+      input = tf.nn.relu(input);
+    }
+    if (clipMax) {
+      Operand<T> lmaxValue = tf.dtypes.cast(tf.constant(maxValue), dataType);
+      Operand<T> zero = tf.dtypes.cast(tf.constant(0), dataType);
+      input = tf.clipByValue(input, zero, lmaxValue);
+    }
+
+    if (alpha != 0.) {
+      input =
+          tf.math.sub(
+              input, tf.math.mul(tf.dtypes.cast(tf.constant(alpha), dataType), negativePart));
```

Comment:
I'm finding it extremely difficult to think through the interactions of all these options. E.g. if we calculate `negativePart` before applying `clipMax`, and then we clip the input, and then we subtract a fraction of `negativePart` from the input, did we do the right thing? How about if there was also a `threshold`?

Perhaps either there's some reasoning we could document up front, or somehow there's a way to order the computation and name the intermediate results so it's clearly correct?

Perhaps the problem is that there are neither documentation, nor examples, nor tests that show how these parameters are supposed to interact.

Or if all of this accurately mimics the Python implementation, we leave it for now while opening a ticket for some newbie like me to come back and tidy it up?

Reply:
This was ported mainly line by line from the Python ReLU method in Keras backend. Java does not have operator overload, so it is a little more cumbersome code than the original python. For example, Python 

`x -= alpha * negative_part`

translates to TF Java as:

```
input =
          tf.math.sub(
              input, tf.math.mul(tf.dtypes.cast(tf.constant(alpha), dataType), negativePart));
```

There may be some benefit to making this in more individual statements, but TF Java in  Graph mode is very difficult to actually debug.  When I have to debug it, I use special code to print the values using a try(Tensor...) logic.

Reply:
My difficulty with the code isn't with this sort of tactic. It's with the overall flow of the computation and the variable names used. Given that it's transliterated from the Python and that the same tests pass, seems good to leave it for another day.